---
title: "Expanded Cell Types - WUI Test (GENCODE Only)"
author: "Mervin M Fansler"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_document: 
    code_folding: show
    toc: true
    toc_float: true
---

# Purpose

Here we compare an expanded set of cell types, with an emphasis on differentiation.

# Initialization

## Libraries
```{r libs, message=FALSE, warning=FALSE}
library(magrittr)
library(tidyverse)
library(cowplot)
library(ggridges)
library(SingleCellExperiment)
library(scutrboot)
library(BiocParallel)
library(plyranges)
library(scuttle)
```

## Parameters
```{r set_params}
set.seed(20211026)
BPPARAM=MulticoreParam(6)
register(BPPARAM)

N_BOOTSTRAPS=10000
MIN_NCELLS=50

GTF_GENCODE="../mca2/data/gff/gencode.vM25.mRNA_ends_found.gff3.gz"
TSV_IPA="../mca2/data/gff/utrome.e30.t5.gc25.pas3.f0.9999.w500.ipa.tsv"
SCE_BRAIN="data/sce/gencode_vM25_pc_w500/ximerakis19.txs.Rds"
SCE_HSPCS="data/sce/gencode_vM25_pc_w500/dahlin18.txs.Rds"
SCE_MESCS="data/sce/gencode_vM25_pc_w500/guo19.txs.Rds"
SCE_TMURIS="data/sce/gencode_vM25_pc_w500/tmuris.txs.Rds"

celltype_map <- c(
  "Bone Marrow (LSK,LK) - 14/Stem"="HSC",
  "Bone Marrow (LSK,LK) - 12/Ery"="Ery",
  "Bone Marrow (LSK,LK) - 7/Neu"="Neutro",
  "Bone Marrow (LSK,LK) - 6/Baso"="Baso",
  "Bone Marrow (LSK,LK) - 5/Mo"="Mono",
  "Brain - OLG"="ODC",
  "Brain - OPC"="OPC",
  "Brain - ImmN"="imNeur",
  "Brain - mNEUR"="mNeur",
  "Trachea - epithelial cell"="TraEpi",
  "Trachea - mesenchymal cell"="TraMes",
  "Mammary_Gland - luminal epithelial cell of mammary gland"="lumMamm",
  "Mammary_Gland - basal cell"="basMamm",
  "Thymus - immature T cell"="Timm",
  "Spleen - T cell"="TSpleen",
  "Spleen - B cell"="BSpleen",
  "Mammary_Gland - T cell"="TMamm",
  "Mammary_Gland - B cell"="BMamm",
  "Brain - ASC"="Astro",
  "Brain - MG"="MG",
  "Brain - NendC"="NeurEC",
  "Brain - EC"="endo",
  "Embryo - mESC"="mESC",
  "Embryo - MEF"="MEF"
)

test_pairs <- list(
  c("mESC", "MEF"),
  c("HSC", "Ery"),
  c("HSC", "Neutro"),
  c("HSC", "Baso"),
  c("HSC", "Mono"),
  c("Timm", "TSpleen"),
  c("OPC", "ODC"),
  c("imNeur", "mNeur"),
  c("TraMes", "TraEpi"),
  c("lumMamm", "basMamm"),
  c("TSpleen", "TMamm"),
  c("TSpleen", "BSpleen"),
  c("mNeur", "NeurEC"),
  c("mNeur", "Astro"),
  c("mNeur", "ODC"),
  c("mNeur", "MG"),
  c("mNeur", "endo")
)
```

# Data
## Load Annotations
```{r}
df_ipa <- read_tsv(TSV_IPA)

df_gc <- read_gff(GTF_GENCODE) %>%
  filter(type == "transcript") %>%
  as_tibble() %>%
  select(transcript_id, gene_name, gene_id, seqnames, start, end, strand) %>%
  mutate(strand=as.character(strand)) %>%
  group_by(gene_id, seqnames) %>%
  mutate(utr_position=ifelse(strand == "+", 
                             rank(end, ties.method="min"), 
                             rank(-start, ties.method="min"))) %>%
  ungroup() %>%
  mutate(is_ipa=transcript_id %in% df_ipa$transcript_id)
```

## Loading
```{r load_data, message=FALSE}
sce_brain <- readRDS(SCE_BRAIN)
colData(sce_brain)['tissue'] <- "Brain"
sce_brain %<>% `[`(, !is.na(.$age))
sce_brain %<>% `[`(, .$age == "young")
colData(sce_brain) %<>% `[`(,c("cell_id", "tissue", "cell_type"))

sce_hspcs <- readRDS(SCE_HSPCS)
colData(sce_hspcs)['tissue'] <- "Bone Marrow (LSK,LK)"
colData(sce_hspcs)['cell_type'] <- sce_hspcs$clusters
colData(sce_hspcs) %<>% `[`(,c("cell_id", "tissue", "cell_type"))

sce_mescs <- readRDS(SCE_MESCS)
colData(sce_mescs)['tissue'] <- "Embryo"
colData(sce_mescs)['cell_type'] <- sce_mescs$sample_id
colData(sce_mescs) %<>% `[`(,c("cell_id", "tissue", "cell_type"))

sce_tmuris <- readRDS(SCE_TMURIS)
colData(sce_tmuris)["cell_type"] <- sce_tmuris$cell_ontology_class
colData(sce_tmuris) %<>% `[`(,c("cell_id", "tissue", "cell_type"))

sce <- cbind(sce_brain, sce_hspcs, sce_mescs, sce_tmuris)

rm(list=c("sce_brain", "sce_hspcs", "sce_mescs", "sce_tmuris"))
gc()
```

## Preprocessing
### Filter SCE
```{r prepare_data}
colData(sce)['tissue_celltype'] <- colData(sce) %>% 
  { str_c(.$tissue, " - ", .$cell_type) }
idx_cells <- colData(sce) %>% { .$tissue_celltype %in% names(celltype_map) }

sce %<>% `[`(,idx_cells)
sce %<>% `[`(rowSums(counts(.)) > 0,)
colData(sce)['celltype_label'] <- celltype_map[sce$tissue_celltype]
gc()
```

### Compute Size Factors
```{r normalize_counts}
sce %<>% computePooledFactors(clusters=.$cell_type, BPPARAM=BPPARAM)
sce %<>% logNormCounts(log=FALSE)
```


### Filter and Annotate Transcripts
```{r tx_annots}
df_rowdata <- rowData(sce) %>% as_tibble %>%
  left_join(df_gc, by=c("transcript_id", "gene_id")) %>%
  filter(!is_ipa) %>%
  add_count(gene_id, name="n_utrs") %>%
  filter(n_utrs > 1) %>%
  group_by(gene_id) %>%
  mutate(utr_rank=rank(utr_position),
         utr_wt=(utr_rank - 1)/(max(utr_rank) - 1)) %>%
  ungroup() %>%
  arrange(seqnames, start) %>%
  dplyr::select(transcript_id, gene_id, gene_name, is_ipa, n_utrs, utr_rank, utr_wt) %>%
  as.data.frame() %>%
  DataFrame(row.names=.$transcript_id)

sce %<>% `[`(df_rowdata$transcript_id,)
rowData(sce) <- df_rowdata

gene_id2sym <- rowData(sce) %>% as_tibble %$% setNames(gene_name, gene_id)
```

# Analysis
## Two-Sample Tests
```{r wd_tests}
res <- bplapply(test_pairs, function (p) {
  testTwoSample(sce, assayName="normcounts",
                sampleKey="celltype_label", geneKey="gene_id",
                sample0=p[[1]], sample1=p[[2]],
                statistic="WUI", featureIndex="utr_wt",
                nBootstraps=N_BOOTSTRAPS, 
                minCellsPerGene=MIN_NCELLS) %>%
    as_tibble() %>%
    mutate(start=p[[1]], end=p[[2]]) %>%
    select(gene, start, end, stat, pval, bootstraps)
})

df_tests <- do.call(rbind, res) %>%
  mutate(gene_symbol=gene_id2sym[gene],
         qval=p.adjust(pval, "BH")) %>%
  select(gene_symbol, gene, everything())

df_tests %>% arrange(pval) %>% head(50) %>% knitr::kable()
```

## P-Values
### All
```{r hist_pvals}
hist(df_tests$pval, breaks=40, xlab="p-value",
     main="Distribution of p-values for Two-Sample WUI Tests")
```

### By Test Pair
```{r hists_by_celltype, fig.width=8, fig.height=12}
df_tests %>%
  mutate(test_pair=str_c(end, " vs ", start)) %>%
  ggplot(aes(x=pval, y=test_pair)) +
  geom_density_ridges(stat='binline', bins=40, scale=0.95) +
  labs(x="p-value", y=NULL, title="WUI Tests") +
  theme_minimal_vgrid()
```

# Export
```{r export}
sprintf("data/dtu/%s-diff-pairwise-wui-tests-gc.min50.boot10k.tsv.gz", format(Sys.time(), '%Y%m%d')) %>%
  write_tsv(x=df_tests)
```


---

# Runtime Details
## Session Info
```{r sesh_info, echo=FALSE}
sessionInfo()
```

## Conda Environment
```{bash conda_env, comment="", echo=FALSE}
if ! command -v conda &> /dev/null
then
  echo "Conda not detected."
elif [ -z "${CONDA_PREFIX}" ]
then
  echo "No active Conda environment."
else
  echo "## Conda Environment YAML"
  conda env export
fi
```
