---
title: "Compare Transcript TPMs - All Batches Downsampled"
author: "Mervin M Fansler"
date: '`r format(Sys.time(), "%d %B %Y")`'
output: 
  html_document: 
    code_folding: show
    toc: true
    toc_float: true
    df_print: paged
---

# Initialization
## Libraries

```{r libs, message=FALSE, warning=FALSE}
library(SummarizedExperiment)
library(SingleCellExperiment)
library(scuttle)
library(Matrix)
library(tidyverse)
library(ggbeeswarm)
library(magrittr)
library(cowplot)

theme_set(theme_bw())
```

## Parameters
```{r params}
set.seed(20231213)

## Sample Names
SAMPLES_BULK <- c('HSCa', 'HSCb', 'HSCc', 'HSCd')
SAMPLES_SC <- str_c("scRNA-seq ", c("1", "2", "3"))
SAMPLES_SC_RAW <- str_c("SIGA", c("B", "C", "D"), "1")
SAMPLES_SC_NAMES <- setNames(SAMPLES_SC, SAMPLES_SC_RAW)
```

## Methods
```{r methods}
compute_column_corrs <- function (df, method="spearman", use="complete.obs") {
  df %>%
    as.matrix %>% cor(method=method, use=use) %>% 
    `[<-`(lower.tri(.,TRUE), NA) %>%
    as_tibble(rownames="sample1") %>% 
    pivot_longer(cols=-1, names_to="sample2", values_to="rho", values_drop_na=TRUE)
}

plot_corrs_all <- function (df, ylab="Spearman Correlation", ylimits=c(0,1)) {
  ggplot(df, aes(x=type, y=rho, fill=type)) +
    stat_summary(fun='mean', geom='bar', width=0.6, color='black') + 
    stat_summary(fun.data=mean_sdl, geom='errorbar', width=0.4) +
    geom_quasirandom(width=0.2, size=0.5) +
    geom_text(aes(x=type, label=str_c("N=", after_stat(count)), y=NULL), 
              stat="count", y=0.03, size=3) +
    scale_y_continuous(expand=expansion(add=c(0,0))) +
    coord_cartesian(ylim=ylimits) +
    labs(x=NULL, y=ylab) +
    guides(fill=FALSE)
}
```

## Load Data
```{r load_data}
se <- readRDS("data/se/hspcs_bulk.txs.rds")

sce <- readRDS("data/sce/hspcs.txs.full_annot.rds") %>%
  `[`(rownames(se),)
if (!("sample_id" %in% colnames(colData(sce))) & ("sample_id.x" %in% colnames(colData(sce)))) {
  colData(sce)['sample_id'] <- sce$sample_id.x
  colData(sce)['sample_id.x'] <- NULL
  colData(sce)['sample_id.y'] <- NULL
}
```

# Data Processing
## Summarize to Samples
```{r sum_samples}
M.cells_sample <- sce$sample_id %>%
  fac2sparse %>% t %>%
  `[`(,SAMPLES_SC_RAW)

cts.sample <- assay(sce, "normcounts") %*% M.cells_sample
colnames(cts.sample) %<>% { SAMPLES_SC_NAMES[.] }

cts.bulk <- assay(se, 'counts') %>%
  `[`(, se$cell_type == 'HSC') %>%
  as.matrix
```

## Downsample Counts
```{r downsample_counts}
min_cts <- min(c(colSums(cts.sample), colSums(cts.bulk)))

prop_ds.sample <- min_cts/colSums(cts.sample)
prop_ds.bulk <- min_cts/colSums(cts.bulk)

cts_ds.sample <- downsampleMatrix(cts.sample, prop=prop_ds.sample, bycol=TRUE)
cts_ds.bulk <- downsampleMatrix(cts.bulk, prop=prop_ds.bulk, bycol=TRUE)
```

### Compare Counts
```{r plt_cts, fig.width=2, fig.height=4}
tibble(
  sample_id=c(colnames(cts.sample), colnames(cts.bulk)),
  cts_raw=c(colSums(cts.sample), colSums(cts.bulk)),
  cts_downsampled=c(colSums(cts_ds.sample), colSums(cts_ds.bulk))
) %>%
  pivot_longer(cols=2:3, names_to="count_type", values_to="counts", names_prefix="cts_") %>%
  ggplot(aes(x=sample_id, y=counts, fill=count_type)) +
  geom_bar(stat='identity', position='identity', color='black', linewidth=0.2) +
  scale_fill_manual(values=c('steelblue', 'grey80'), guide=guide_none()) +
  scale_x_discrete(guide=guide_axis(angle=90)) +
  scale_y_continuous(expand=c(0,0,0.1,0)) +
  labs(x=NULL, y="Counts", fill="Type")
```


## Compute TPMs
```{r compute_tpm}
tpm.hsc.sc <- cts_ds.sample %>% 
  { . %*% Diagonal(1e6/colSums(.), n=ncol(.), names=TRUE) } %>%
  as.matrix

tpm.hsc.bulk <- cts_ds.bulk %>% 
  { . %*% Diagonal(1e6/colSums(.), n=ncol(.), names=TRUE) } %>%
  as.matrix

all(rownames(tpm.hsc.bulk) == rownames(tpm.hsc.sc)) %>%
  { if (.) "Yes" else "No"} %>%
  sprintf(fmt="Are all genes present and in identical order? %s")

## convert to tibbles, then join
df_bulk <- as_tibble(tpm.hsc.bulk, rownames="transcript_id")
df_sc <- as_tibble(tpm.hsc.sc, rownames="transcript_id")

df_tpm <- left_join(df_bulk, df_sc, by="transcript_id") %>%
  right_join(x=rowData(se)[,c("transcript_id", "gene_id")] %>% as.data.frame, by="transcript_id")
```


# LUI
```{r compute_lui}
SAMPLE_COLS=c(SAMPLES_BULK, SAMPLES_SC)

df_annots <- rowData(sce) %>% as_tibble %>%
  select(transcript_id, utr_position)

df_lui <- df_tpm %>% 
  left_join(df_annots, by='transcript_id') %>%
  pivot_longer(all_of(SAMPLE_COLS), names_to='sample', values_to="tpm") %>%
  group_by(gene_id, sample) %>%
  filter(dplyr::n() > 1) %>%
  mutate(lui=tpm/sum(tpm, na.rm=TRUE)) %>%
  filter(utr_position == max(utr_position)) %>% 
  ungroup() %>%
  select(gene_id, sample, lui) %>%
  pivot_wider(names_from='sample', values_from='lui')

MIN_TPM=0
df_lui_two <- df_tpm %>% 
  left_join(df_annots, by='transcript_id') %>%
  pivot_longer(all_of(SAMPLE_COLS), names_to='sample', values_to="tpm") %>%
  group_by(transcript_id) %>%
  filter(all(tpm > MIN_TPM)) %>%
  ungroup() %>%
  group_by(gene_id, sample) %>%
  filter(dplyr::n() == 2) %>%
  mutate(lui=tpm/sum(tpm, na.rm=TRUE)) %>%
  filter(utr_position == max(utr_position)) %>% 
  ungroup() %>%
  select(gene_id, sample, lui) %>%
  pivot_wider(names_from='sample', values_from='lui')
```


## Combined
### TPM
```{r compare_tpm_combined, fig.width=3, fig.height=4, warning=FALSE}
SAMPLE_COLS_TPM=c(SAMPLES_BULK, SAMPLES_SC)

df_tpm_corrs <- df_tpm[,SAMPLE_COLS_TPM] %>% 
  compute_column_corrs %>%
  mutate(type=str_c(ifelse(str_detect(sample1, "scRNA"), "sc", "bulk"),
               ifelse(str_detect(sample2, "scRNA"), "sc", "bulk"),
               sep="-"))

plot_corrs_all(df_tpm_corrs, ylab="Batch-Batch TPM Spearman Correlation")

# ggsave("output/figure2/ed2a-sommerkamp-tpm-corrs-bars-wholesample.pdf",
#        width=3, height=4, dpi=300)
```

```{r compare_tpm_combined_zoom, fig.width=3, fig.height=4, warning=FALSE}
plot_corrs_all(df_tpm_corrs, ylab="Batch-Batch TPM Spearman Correlation", ylimits=c(0.8, 1))

# ggsave("output/figure2/ed2a-sommerkamp-tpm-corrs-bars-zoom-wholesample.pdf",
#        width=3, height=4, dpi=300)
```


### LUI - All Multi-UTR
```{r compare_lui_combined, fig.width=3, fig.height=4, warning=FALSE}
SAMPLE_COLS_LUI=c(SAMPLES_BULK, SAMPLES_SC)

df_lui_corrs <- df_lui[,SAMPLE_COLS_LUI] %>% 
  compute_column_corrs %>%
  mutate(type=str_c(ifelse(str_detect(sample1, "scRNA"), "sc", "bulk"),
               ifelse(str_detect(sample2, "scRNA"), "sc", "bulk"),
               sep="-"))

plot_corrs_all(df_lui_corrs, ylab="Batch-Batch LUI Spearman Correlation")


# ggsave("output/figure2/ed2b-sommerkamp-lui-multiutr-corrs-bars-wholesample.pdf",
#        width=3, height=4, dpi=300)
```

### LUI - Two-UTR
```{r compare_lui_twoutr_combined, fig.width=3, fig.height=4, warning=FALSE}
SAMPLE_COLS_LUI=c(SAMPLES_BULK, SAMPLES_SC)

df_lui_two_corrs <- df_lui_two[,SAMPLE_COLS_LUI] %>% 
  compute_column_corrs %>%
  mutate(type=str_c(ifelse(str_detect(sample1, "scRNA"), "sc", "bulk"),
               ifelse(str_detect(sample2, "scRNA"), "sc", "bulk"),
               sep="-"))

plot_corrs_all(df_lui_two_corrs, ylab="Batch-Batch LUI Spearman Correlation")

# ggsave("output/figure2/ed2b-sommerkamp-lui-twoutr-corrs-bars-wholesample.pdf",
#        width=3, height=4, dpi=300)
```


# Tables
```{r table}
full_join(df_tpm_corrs, df_lui_two_corrs, by=c("sample1", "sample2", "type"),
          suffix=c(x="_TPM", y="_LUI")) %>%
  select(sample1, sample2, rho_TPM, rho_LUI, type) %>%
  arrange(type, sample1, sample2) %T>%
  write_tsv("output/tables/tbl-hspcs-bulk-sc-corrs-wholesample-downsampled.tsv")
```


---

# Runtime Details
## Session Info
```{r sesh_info, echo=FALSE}
sessionInfo()
```

## Conda Environment
```{bash comment="", echo=FALSE}
if ! command -v conda &> /dev/null
then
  echo "Conda not detected."
elif [ -z "${CONDA_PREFIX}" ]
then
  echo "No active Conda environment."
else
  echo "## Conda Environment YAML"
  conda env export
fi
```
