---
title: "Compare Transcript TPMs - Batches"
author: "Mervin M Fansler"
date: '`r format(Sys.time(), "%d %B %Y")`'
output: 
  html_document: 
    code_folding: show
    toc: true
    toc_float: true
---

# Libraries

```{r libs, message=FALSE, warning=FALSE}
library(SummarizedExperiment)
library(SingleCellExperiment)
library(Matrix)
library(tidyverse)
library(magrittr)
library(cowplot)
library(GGally)
library(ggrastr)

theme_set(theme_bw())
```

# Methods
```{r methods}
lower_geom_point_pct <- function (data, mapping, ...) {
  ggplot(data=data, mapping=mapping, ...) +
    geom_point_rast(size=0.2, alpha=0.2, pch=16, ...) +
    geom_abline(linetype='dashed', color='red') +
    lims(x=c(0,1), y=c(0,1))
}

lower_geom_point <- function (data, mapping, ...) {
  ggplot(data=data, mapping=mapping, ...) +
    geom_point_rast(size=0.2, alpha=0.2, pch=16, ...) +
    geom_abline(linetype='dashed', color='red')
}
```

# Load Data
```{r load_data}
se <- readRDS("data/se/hspcs_bulk.txs.rds")

sce <- readRDS("data/sce/hspcs.txs.full_annot.rds") %>%
  `[`(rownames(se),)
if (!("sample_id" %in% colnames(colData(sce))) & ("sample_id.x" %in% colnames(colData(sce)))) {
  colData(sce)['sample_id'] <- sce$sample_id.x
  colData(sce)['sample_id.x'] <- NULL
  colData(sce)['sample_id.y'] <- NULL
}
```

## Compute TPM
```{r compute_tpm}
## Sample Names
SAMPLES_BULK <- c('HSCb', 'HSCc')
SAMPLES_SC_14 <- str_c("14/Stem_SIGA", c("B", "C"), "1")
SAMPLES_SC_0 <- str_c("0_SIGA", c("B", "C"), "1")
SAMPLES_SC_COMBINED <- str_c("scRNA-seq ", c("1", "2"))
SAMPLES_SC <- c(SAMPLES_SC_14, SAMPLES_SC_0)
SAMPLES_SC_ALL <- c(SAMPLES_SC, SAMPLES_SC_COMBINED)

M.cells_celltype_sample <- colData(sce) %>% 
  as_tibble %$%
  str_c(clusters, sample_id, sep="_") %>%
  fac2sparse %>% t %>%
  `[`(,SAMPLES_SC) %>%
  cbind(.[,SAMPLES_SC_14] + .[,SAMPLES_SC_0]) %>%
  `colnames<-`(SAMPLES_SC_ALL)

cts.celltype_sample <- assay(sce, "normcounts") %*% M.cells_celltype_sample

tpm.hsc.sc <- cts.celltype_sample %>%
  { . %*% Diagonal(1e6/colSums(.), n=ncol(.), names=TRUE) } %>%
  as.matrix

tpm.hsc.bulk <- assay(se, 'counts') %>% 
  { . %*% Diagonal(1e6/colSums(.), n=ncol(se), names=TRUE) } %>%
  `[`(, se$cell_type == 'HSC') %>%
  as.matrix

all(rownames(tpm.hsc.bulk) == rownames(tpm.hsc.sc)) %>%
  { if (.) "Yes" else "No"} %>%
  sprintf(fmt="Are all genes present and in identical order? %s")

## convert to tibbles, then join
df_bulk <- as_tibble(tpm.hsc.bulk, rownames="transcript_id")
df_sc <- as_tibble(tpm.hsc.sc, rownames="transcript_id")

df_tpm <- left_join(df_bulk, df_sc, by="transcript_id") %>%
  right_join(x=rowData(se)[,c("transcript_id", "gene_id")] %>% as.data.frame, by="transcript_id")
```


# LUI
```{r compute_lui}
SAMPLE_COLS=c(SAMPLES_BULK, SAMPLES_SC_ALL)

df_annots <- rowData(sce) %>% as_tibble %>%
  select(transcript_id, utr_position)

df_lui <- df_tpm %>% 
  left_join(df_annots, by='transcript_id') %>%
  pivot_longer(all_of(SAMPLE_COLS), names_to='sample', values_to="tpm") %>%
  group_by(gene_id, sample) %>%
  filter(n() > 1) %>%
  mutate(lui=tpm/sum(tpm, na.rm=TRUE)) %>%
  filter(utr_position == max(utr_position)) %>% 
  ungroup() %>%
  select(gene_id, sample, lui) %>%
  pivot_wider(names_from='sample', values_from='lui')

MIN_TPM=0
df_lui_two <- df_tpm %>% 
  left_join(df_annots, by='transcript_id') %>%
  pivot_longer(all_of(SAMPLE_COLS), names_to='sample', values_to="tpm") %>%
  group_by(transcript_id) %>%
  filter(all(tpm > MIN_TPM)) %>%
  ungroup() %>%
  group_by(gene_id, sample) %>%
  filter(n() == 2) %>%
  mutate(lui=tpm/sum(tpm, na.rm=TRUE)) %>%
  filter(utr_position == max(utr_position)) %>% 
  ungroup() %>%
  select(gene_id, sample, lui) %>%
  pivot_wider(names_from='sample', values_from='lui')
```


## Combined
### TPM
```{r compare_tpm_combined, fig.width=6, fig.height=6, warning=FALSE}
SAMPLE_COLS_TPM=c(SAMPLES_BULK, SAMPLES_SC_COMBINED)

# all genes
df_tpm %>%
  mutate(across(all_of(SAMPLE_COLS_TPM), ~ log10(.+1))) %>%
  ggpairs(columns=SAMPLE_COLS_TPM, progress=FALSE,
          lower=list(continuous=wrap(lower_geom_point, alpha=0.4)),
          upper=list(continuous=wrap('cor', method="spearman"))) +
  labs(x="log10(1 + TPM)", y="log10(1 + TPM)")

ggsave("output/figure2/ed2e-sommerkamp-all-tpm.pdf", width=6, height=6, dpi=300)
```

### LUI - All Multi-UTR
```{r compare_lui_combined, fig.width=6, fig.height=6, warning=FALSE}
SAMPLE_COLS_LUI=c(SAMPLES_BULK, SAMPLES_SC_COMBINED)

# all multi-UTR genes in UTRome
df_lui %>%
  ggpairs(columns=SAMPLE_COLS_LUI, progress=FALSE,
          lower=list(continuous=wrap(lower_geom_point_pct, alpha=0.4)),
          upper=list(continuous=wrap('cor', method="spearman"))) +
  labs(x="LUI", y="LUI")
```

### LUI - Two-UTR
```{r compare_twolui_combined, fig.width=6, fig.height=6, warning=FALSE}
df_lui_two %>%
  ggpairs(columns=SAMPLE_COLS_LUI, progress=FALSE,
          lower=list(continuous=wrap(lower_geom_point_pct, alpha=0.4)),
          upper=list(continuous=wrap('cor', method="spearman"))) +
  labs(x="LUI", y="LUI")

ggsave("output/figure2/ed2g-sommerkamp-all-lui-two-utr.pdf", width=6, height=6, dpi=300)
```

# Exported Figures
## Bulk Compare
### TPM
```{r fig_bulk_tpm, fig.width=5, fig.height=4}
axis_lims <- c(df_tpm$HSCb, df_tpm$HSCc) %>%
  { log10(1 + .) } %>% { c(min(.), max(.)) } 
label_lims <- axis_lims %>% { . + c(0.25*(.[2]-.[1]), 0.05*(.[1]-.[2])) }
label_rho_s <- sprintf("\"Spearman\" ~ rho == \"%0.3f\"", 
                       cor(df_tpm$HSCb, df_tpm$HSCc, method='spearman'))

df_tpm %>% 
  ggplot(aes(x=log10(HSCb+1), y=log10(HSCc+1))) +
  geom_point_rast(size=0.2, alpha=0.3, pch=16) +
  geom_abline(linetype='dashed', color='red') +
  annotate("text", x=label_lims[1], y=label_lims[2], label=label_rho_s, 
           size=4, parse=TRUE) +
  scale_x_continuous(limits=axis_lims) + scale_y_continuous(limits=axis_lims) +
  labs(x="Bulk 3'-seq 1 [log10(TPM + 1)]", y="Bulk 3'-seq 2 [log10(TPM + 1)]") +
  theme(plot.margin=grid::unit(c(4,8,4,4), "mm"), aspect.ratio=1)

ggsave("output/figure2/fig2a-sommerkamp-bulk-bulk-tpm.pdf", width=5, height=4)
```

```{r fig_bulk_lui, fig.width=5, fig.height=4}
label_rho_s <- sprintf("\"Spearman\" ~ rho == \"%0.3f\"", 
                       cor(df_lui_two$HSCb, df_lui_two$HSCc, method='spearman'))

df_lui_two %>% 
  ggplot(aes(x=HSCb, y=HSCc)) +
  geom_point_rast(size=0.3, alpha=0.4, pch=16) +
  geom_abline(linetype='dashed', color='red') +
  annotate("text", x=0.25, y=0.95, label=label_rho_s, size=4, parse=TRUE) +
  scale_x_continuous(limits=c(0,1), expand=c(0,0)) +
  scale_y_continuous(limits=c(0,1), expand=c(0,0)) +
  labs(x="Bulk 3'-seq 1 [LUI]", y="Bulk 3'-seq 2 [LUI]") +
  theme(plot.margin=grid::unit(c(4,8,4,4), "mm"), aspect.ratio=1)

ggsave("output/figure2/fig2d-sommerkamp-bulk-bulk-lui-two-utr.pdf", width=5, height=4)
```

## Bulk vs scRNA-seq
### TPM
```{r fig_compare_tpm, fig.width=5, fig.height=4}
axis_lims <- c(df_tpm$HSCb, df_tpm$`scRNA-seq 1`) %>%
  { log10(1 + .) } %>% { c(min(.), max(.)) } 
label_lims <- axis_lims %>% { . + c(0.25*(.[2]-.[1]), 0.05*(.[1]-.[2])) }
label_rho_s <- sprintf("\"Spearman\" ~ rho == \"%0.3f\"", 
                       cor(df_tpm$HSCb, df_tpm$`scRNA-seq 1`, method='spearman'))

df_tpm %>% 
  ggplot(aes(x=log10(HSCb+1), y=log10(`scRNA-seq 1`+1))) +
  geom_point_rast(size=0.2, alpha=0.3, pch=16) +
  geom_abline(linetype='dashed', color='red') +
  annotate("text", x=label_lims[1], y=label_lims[2], label=label_rho_s, 
           size=4, parse=TRUE) +
  scale_x_continuous(limits=axis_lims) + scale_y_continuous(limits=axis_lims) +
  labs(x="Bulk 3'-Seq 1 [log10(TPM + 1)]", y="scRNA-seq 1 [log10(TPM + 1)]") +
  theme(plot.margin=grid::unit(c(4,8,4,4), "mm"), aspect.ratio=1)

ggsave("output/figure2/fig2b-sommerkamp-bulk-sc-tpm.pdf", width=5, height=4)
```

```{r fig_compare_lui, fig.width=5, fig.height=4}
label_rho_s <- sprintf("\"Spearman\" ~ rho == \"%0.3f\"", 
                       cor(df_lui_two$HSCb, df_lui_two$`scRNA-seq 1`, method='spearman'))

df_lui_two %>% 
  ggplot(aes(x=HSCb, y=`scRNA-seq 1`)) +
  geom_point_rast(size=0.3, alpha=0.4, pch=16) +
  geom_abline(linetype='dashed', color='red') +
  annotate("text", x=0.25, y=0.95, label=label_rho_s, size=4, parse=TRUE) +
  scale_x_continuous(limits=c(0,1), expand=c(0,0)) +
  scale_y_continuous(limits=c(0,1), expand=c(0,0)) +
  labs(x="Bulk 3'-seq 1 [LUI]", y="scRNA-seq 1 [LUI]") +
  theme(plot.margin=grid::unit(c(4,8,4,4), "mm"), aspect.ratio=1)

ggsave("output/figure2/fig2e-sommerkamp-bulk-sc-lui-two-utr.pdf", width=5, height=4)
```

## scRNA-seq Compare
### TPM
```{r fig_sc_tpm, fig.width=5, fig.height=4}
axis_lims <- c(df_tpm$`scRNA-seq 2`, df_tpm$`scRNA-seq 1`) %>%
  { log10(1 + .) } %>% { c(min(.), max(.)) } 
label_lims <- axis_lims %>% { . + c(0.25*(.[2]-.[1]), 0.05*(.[1]-.[2])) }
label_rho_s <- sprintf("\"Spearman\" ~ rho == \"%0.3f\"", 
                       cor(df_tpm$`scRNA-seq 1`, df_tpm$`scRNA-seq 2`, method='spearman'))

df_tpm %>% 
  ggplot(aes(x=log10(`scRNA-seq 1`+1), y=log10(`scRNA-seq 2`+1))) +
  geom_point_rast(size=0.2, alpha=0.3, pch=16) +
  geom_abline(linetype='dashed', color='red') +
  annotate("text", x=label_lims[1], y=label_lims[2], label=label_rho_s, 
           size=4, parse=TRUE) +
  scale_x_continuous(limits=axis_lims) + scale_y_continuous(limits=axis_lims) +
  labs(x="scRNA-seq 1 [log10(TPM + 1)]", y="scRNA-seq 2 [log10(TPM + 1)]") +
  theme(plot.margin=grid::unit(c(4,8,4,4), "mm"), aspect.ratio=1)

ggsave("output/figure2/fig2c-sommerkamp-sc-sc-tpm.pdf", width=5, height=4)
```

```{r fig_sc_lui, fig.width=5, fig.height=4}
label_rho_s <- sprintf("\"Spearman\" ~ rho == \"%0.3f\"", 
                       cor(df_lui_two$`scRNA-seq 1`, df_lui_two$`scRNA-seq 2`, method='spearman'))

df_lui_two %>% 
  ggplot(aes(x=`scRNA-seq 1`, y=`scRNA-seq 2`)) +
  geom_point_rast(size=0.3, alpha=0.4, pch=16) +
  geom_abline(linetype='dashed', color='red') +
  annotate("text", x=0.25, y=0.95, label=label_rho_s, size=4, parse=TRUE) +
  scale_x_continuous(limits=c(0,1), expand=c(0,0)) +
  scale_y_continuous(limits=c(0,1), expand=c(0,0)) +
  labs(x="scRNA-seq 1 [LUI]", y="scRNA-seq 2 [LUI]") +
  theme(plot.margin=grid::unit(c(4,8,4,4), "mm"), aspect.ratio=1)

ggsave("output/figure2/fig2f-sommerkamp-sc-sc-lui-two-utr.pdf", width=5, height=4)
```

---

# Runtime Details
## Session Info
```{r sesh_info, echo=FALSE}
sessionInfo()
```

## Conda Environment
```{bash comment="", echo=FALSE}
if ! command -v conda &> /dev/null
then
  echo "Conda not detected."
elif [ -z "${CONDA_PREFIX}" ]
then
  echo "No active Conda environment."
else
  echo "## Conda Environment YAML"
  conda env export
fi
```
