---
title: "Melanoma Resistance (451Lu) - DGE vs DTU"
author: "Mervin Fansler"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_document: 
    code_folding: show
    toc: true
    toc_float: true
---

# Purpose

Comparison of significant genes in either gene expression or 3' UTR isoform usage.

# Initialization

## Libraries
```{r libs, message=FALSE, warning=FALSE}
library(magrittr)
library(tidyverse)
library(cowplot)
library(writexl)
library(org.Hs.eg.db)
```

## Parameters
```{r set_params}
MAX_QVAL = 0.05
MIN_DLUI = 0.15
MIN_L2FC = log2(1.5)
```


# Data
## Loading
```{r load_data, message=FALSE}
df_test_lui <- read_tsv("data/dtu/20211110-451lu-twosample-lui.min50.boot10k.tsv.gz")
df_test_dge <- read_tsv("data/dge/20211110-451lu-t-tests.tsv.gz")
```

# Analysis

## LUI Test: 451Lu iBRAF Resistant vs Parental
### Significant Genes
```{r sig_lui}
df_test_lui %>%
  filter(qval < MAX_QVAL, abs(stat) >= MIN_DLUI) %>%
  mutate(direction=ifelse(stat > 0, "lengthening", "shortening")) %>%
  dplyr::count(direction) %>%
  knitr::kable()
```


## Compare DGE
```{r sig_dge}
df_test_dge %>%
  filter(qval < MAX_QVAL, abs(log2FC) >= MIN_L2FC) %>%
  mutate(direction=ifelse(log2FC > 0, "up", "down")) %>%
  dplyr::count(direction) %>%
  knitr::kable()
```

```{r compare_dge}
df_sig_lui <- df_test_lui %>%
  mutate(sig_lui=qval < MAX_QVAL & abs(stat) >= MIN_DLUI) %>%
  dplyr::select(gene, sig_lui)

df_sig_dge <- df_test_dge %>%
  mutate(sig_gene=qval < MAX_QVAL & abs(log2FC) >= MIN_L2FC) %>%
  dplyr::rename(gene=gene_id) %>%
  dplyr::select(gene, sig_gene)

df_sig_multi <- df_sig_lui %>%
  left_join(df_sig_dge, by="gene") %>%
  replace_na(list(sig_gene=FALSE))

df_sig_multi %$%
  table(sig_gene, sig_lui) %T>%
  print() %>%
  chisq.test(simulate.p.value=TRUE, B=10000)
```

# Export LUI Table
```{r export}
df_test_lui_final <- df_test_lui %>%
  dplyr::rename(gene_id=gene, dLUI=stat) %>%
  left_join(dplyr::select(df_test_dge, gene_id, log2FC, pval, qval), by="gene_id",
            suffix=c(x="_dLUI", y="_L2FC")) %>%
  mutate(ensembl_id=str_extract(gene_id, "^[^.]+"))

mapIds(org.Hs.eg.db, df_test_lui_final$ensembl_id, "SYMBOL", "ENSEMBL", multiVals="first") %>%
  enframe(name="ensembl_id", value="gene_symbol") %>%
  left_join(x=df_test_lui_final, by="ensembl_id") %>%
  dplyr::select(gene_id, gene_symbol, 
                dLUI, pval_dLUI, qval_dLUI, bootstraps,
                log2FC, pval_L2FC, qval_L2FC) %>%
  arrange(pval_dLUI, -abs(dLUI)) %T>%
  write_xlsx("output/tables/tbl4-451Lu-resistant-lui-test-t-test.bs10k.xlsx") %>%
  head(n=100) %>%
  knitr::kable()
```



---

# Runtime Details
## Session Info
```{r sesh_info, echo=FALSE}
sessionInfo()
```

## Conda Environment
```{bash conda_env, comment="", echo=FALSE}
if ! command -v conda &> /dev/null
then
  echo "Conda not detected."
elif [ -z "${CONDA_PREFIX}" ]
then
  echo "No active Conda environment."
else
  echo "## Conda Environment YAML"
  conda env export
fi
```
