---
title: "Fansler et al., Nature Communications, Figure 5"
author: "Sibylle Mitschka"
output: 
  html_document: 
    code_folding: show
    df_print: paged
---

## Loading of R packages
(not all may be required for this code section)

```{r message=FALSE, warning=FALSE}

library(tidyverse)
library(broom)
library(Biostrings)
library(ggpubr)
library(rtracklayer)
library(org.Hs.eg.db)
library(RColorBrewer)
library(biomaRt)
library(GenomicFeatures)
library(BSgenome)
library(GeneOverlap)
library(pheatmap)
library(biomaRt)
library(plyranges)
library(ggrepel)
library(writexl)
library(gridExtra)
library(readxl)

```

## Introduction

The data for this analysis have been generated by MMF via 3'UTR isoform-specific quantification with the scUTRquant tool using the "human UTRome". The used data are from a CRISPRi screen performed in human K562 cells published by Replogle et al Cell 2022. In total, this screen comprised 2.134 targeting perturbations and 97 non-targeting control sgRNA samples. A clustering of z-scaled dWUI values using a subset of response genes generated a total of 18 perturbation clusters comprising 836 genes.

In this analysis, we generate figures describing the effect of these perturbation clusters with respect to the number of regulated genes (APA only), their effect direction (lengthening vs shortening) and effect strength using the dWUI metric (Fig 5a). 
We perform pairwise comparisons of the sets of regulated genes in order to explore cooperative or antagonistic regulation of multi-UTR genes by different clusters (Fig 5b and 5c). 
Furthermore, we determined whether APA isoform changes are balanced (compensated), or not (Fig 5d). 
Finally, we use correlation analysis to determine which gene or transcript features are associated with APA regulation in each cluster (see also Suppl Figure 8).


## Loading relevant data files
```{r message=FALSE, warning=FALSE}

### this data file comprises the gene expression counts and WUI and pct_ipa values for all sgRNA and gene pair conditions

df_WUIess <-readRDS("input_data/df_wui_kd6_essential_ui10.Rds")

### This file provides the cluster identity for a subset of the perturbations (836 genes) which will be used to summarize the data on the cluster level, the clsuter_ids provided in this file do not match the final cluster numbers presented in the paper

target_clus <- read.csv("input_data/df_target_nnclusters_kd6_essential_ui10.csv") %>% mutate(cluster_id_target= as.factor(cluster_id))

### This file provides the names for the clusters and the conversion of the primary cluster id to the final cluster id used in the paper.

names <- read.csv("input_data/names_perturbation_cluster.csv") %>% 
  mutate(cluster_id_target = as.factor(cluster_id_target))

### This file provides additional data on the subset of mRNA 3'ends that have been detected in this K562 data set. They will be used to derive further features such as mRNA lengths from the data.

anno_ess <- read.csv("input_data/df_utrs_kd6_essential_ui10.csv")

```



```{r}

nontarg_ess <- df_WUIess %>% 
  filter(target_gene=="non-targeting") %>% 
  group_by(gene_name, gene_id) %>%
  summarize(mean_tpm = mean(tpm, na.rm=TRUE),
            sd_tpm = sd(tpm, na.rm=TRUE),
            mean_wui = mean(wui, na.rm=TRUE),
            sd_wui= sd(wui, na.rm=TRUE), 
            mean_ipa = mean(pct_ipa, na.rm=TRUE),
            sd_ipa= sd(pct_ipa, na.rm=TRUE)) %>%
  mutate(mean_wui = if_else(mean_tpm>= 5, mean_wui, NA),
         mean_ipa = if_else(mean_tpm>= 5 & mean_ipa>=0.1 & mean_ipa<= 0.9 , mean_ipa, NA),
         sd_wui= if_else(mean_tpm>= 5,sd_wui, NA),
         sd_ipa= if_else(mean_tpm>= 5 & mean_ipa>=0.1 & mean_ipa<= 0.9 ,sd_ipa, NA))

            
df_WUIess1 <- left_join(df_WUIess, 
                        nontarg_ess, by=c("gene_id", "gene_name"), 
                        multiple="all") %>%
              mutate(dwui = wui-mean_wui, 
                     dipa = pct_ipa-mean_ipa,
                     fc_tpm = log2(tpm/mean_tpm))

df_WUI_ess_clus <- left_join(df_WUIess1, target_clus, by=c("target_gene", "target_gene_id", "sgID_AB"))

df_WUI_ess_clus_sum <-  df_WUI_ess_clus %>% 
  dplyr::filter(!is.na(wui)) %>% 
  group_by(gene_name, gene_id, cluster_id_target) %>% 
  summarize(mean_dwui= mean(dwui, na.rm=T))

### loading of DUL test summary table and merging with table

DUL_test <- read_delim("input_data/df_dwui_kd6_clusters.mw_test.tsv.gz") %>% 
  mutate(cluster_id_target=as.factor(cluster_id), 
  DUL = as.factor(case_when(p.adj < 0.05 & dwui < 0 ~ "Shortening", 
                            p.adj < 0.05 & dwui > 0 ~ "Lengthening",
                            .default= "No change"))) %>%
  dplyr::select(gene_name, gene_id, cluster_id_target, DUL, p.adj)

df_WUI_ess_clus_sum1 <- left_join(df_WUI_ess_clus_sum, DUL_test, by= c("gene_name", "gene_id", "cluster_id_target")) 
```


```{r}

#After significance testing on the cluster-level, we created overview plots for the number of significant genes per cluster and their effect size and direction.

overview <- df_WUI_ess_clus_sum1 %>% 
  filter(!DUL == "No change", !cluster_id_target %in% c("control", "NA")) %>% 
  group_by(cluster_id_target) %>% 
  dplyr::count(DUL)  %>% 
  left_join(names, by="cluster_id_target")

Figure5_a1 <- overview %>% mutate(n = ifelse(DUL=="Shortening", -1*n, n)) %>% 
  ggplot(aes(x= reorder(clus_name, - order_heatm_new), y= n, fill=DUL, label = abs(n))) + 
        geom_bar(stat = "identity") +
        coord_flip()+
        xlab("") + 
        geom_text(size = 3, position = position_stack(vjust = 0.5))+
        scale_y_continuous("Genes")+
        scale_fill_manual(values = c("#ef4136","#27aae1"))+
        theme_bw()

Figure5_a1

ggsave("Figure_outputs/Fig5a1.pdf", Figure5_a1, height=6, width=6)

df_WUI_ess_clus_sum2 <- df_WUI_ess_clus_sum1 %>% 
  ungroup() %>%
  filter(!DUL == "No change", !cluster_id_target %in% c("control", "NA")) %>% 
  left_join(names, by="cluster_id_target") %>% ungroup()

Figure5_a2 <- df_WUI_ess_clus_sum2 %>% 
  dplyr::filter(!cluster_id_target == "control", !is.na(cluster_id_target), 
         DUL %in% c("Shortening","Lengthening")) %>%
  ggplot(aes(x= mean_dwui, y= reorder(clus_name, - order_heatm_new, ), color=DUL)) + 
  geom_jitter(alpha=0.3, width = 0, height = 0.3, size = 1) +
  xlab("dWUI") + 
  ylab("")+
  scale_color_manual(values = c("#ef4136","#27aae1"))+
  theme_bw()

Figure5_a2

ggsave("Figure_outputs/Fig5a2.pdf", Figure5_a2, height=6, width=6)

```


Next, we investigate the overlaps of the sets of regulated genes in each cluster, distinguishing between shortening and lengthening genes. For this analysis, we use the GeneOverlaps R package to perform all pair-wise comparisons. 


```{r}

### Create gene lists by cluster for shortening and lengthening. For the object containing list with lengthening genes, the cluster for Chromatin modifiers needs to be manually added as an empty list (no genes contained).
### Cluster IDs match the nomenclature used in the paper (called order_heatm_new).

clus_apa_sho <-df_WUI_ess_clus_sum2 %>% 
  filter(DUL %in% c("Shortening")) %>% 
  group_by(order_heatm_new) %>%
  summarize(hits= list(gene_name)) %>%  
  deframe()

clus_apa_len <-df_WUI_ess_clus_sum2 %>% 
  filter(DUL %in% c("Lengthening")) %>% 
  group_by(order_heatm_new) %>%
  summarize(hits= list(gene_name)) %>% 
  deframe() %>% 
  append("",after=4) 

## An empty list for cluster five and lengthening genes was manually added.

names(clus_apa_len)[[5]] <- as.character(5)

```

## Construct gene overlap objects and create plots

For statistical testing, we need to define the size of the overall gene universe that should be considered for random sets of equal size. For this purpose, we decided to only consider genes that have been found to be significant for APA in at least one cluster (2,092). This is a conservative assumption.
For each "geneOverlap" object, we are calculating -log10-transformed p.values (Fisher's exact test) and create a comparison matrix. We are substituting self-comparisons along the diagonal axis with NA values.

```{r}

### Calculate the gene set size of the multi-UTR gene universe

genes_apa_conservative <- df_WUI_ess_clus_sum2 %>% 
  dplyr::count(gene_name) %>% 
  pull(gene_name)

gs_apa_conservative <- length(genes_apa_conservative)

### This defines the breaks used in the color scale of two of the heatmaps

breaksList = seq(0, 50, by = 5)

### Shortening gene sets

gom.obj_sho <-newGOM(clus_apa_sho, clus_apa_sho, genome.size=gs_apa_conservative)

clus_apa_sho_matrix_pval <- -log10(getMatrix(gom.obj_sho, name="pval"))

diag(clus_apa_sho_matrix_pval) <- NA

clus_apa_sho_matrix_pval_plot <- 
  pheatmap(clus_apa_sho_matrix_pval, 
           cluster_rows = F, 
           cluster_cols = F, 
           display_numbers = FALSE, 
           col = colorRampPalette(brewer.pal(8, "Blues"))(10),
           breaks = breaksList)

clus_apa_sho_matrix_pval_plot

ggsave("Figure_outputs/Fig5b_sho.pdf", clus_apa_sho_matrix_pval_plot)

### Lengthening gene sets

gom.obj_len <-newGOM(clus_apa_len, clus_apa_len, genome.size=gs_apa_conservative)

clus_apa_len_matrix_pval <- -log10(getMatrix(gom.obj_len, name="pval"))  

diag(clus_apa_len_matrix_pval) <- NA

clus_apa_len_matrix_pval_plot <- 
  pheatmap(clus_apa_len_matrix_pval, 
           cluster_rows = F, 
           cluster_cols = F, 
           display_numbers = FALSE, 
           col=colorRampPalette(brewer.pal(8, "Reds"))(10), 
           breaks=breaksList)

clus_apa_len_matrix_pval_plot

ggsave("Figure_outputs/Fig5b_len.pdf", clus_apa_len_matrix_pval_plot)

### Shortening vs lengthening sets in each clusters

gom.obj_apa_cross <-newGOM(clus_apa_sho, clus_apa_len, genome.size=gs_apa_conservative)

apa_cross_pval <- -log10(getMatrix(gom.obj_apa_cross, name="pval")) 

apa_cross_pval[sapply(apa_cross_pval, is.infinite)] <- NA

apa_cross_pval_plot <- 
  pheatmap(apa_cross_pval, 
           cluster_rows = F, 
           cluster_cols = F, 
           display_numbers = FALSE, 
           col=colorRampPalette(brewer.pal(8, "BuPu"))(10))

apa_cross_pval_plot

ggsave("Figure_outputs/Fig5c.pdf", apa_cross_pval_plot)

```

Next, we assess what fraction of significantly regulated genes in each cluster exhibit balanced isoform expression changes, in contrast to a uni-directional regulation (i.e, only one isoform is regulated). For simplicity, we only study multi-UTR genes with exactly two 3'UTR isoforms (3,489 genes). We define balanced isoform regulation as an event where isoforms are oppositely regulated, with a minimum of 50% of the expression being shifted to the other isoform. 


```{r}

# Identify genes with exactly two 3'UTR isoforms in the data set

anno_ess_2UTR_list <- anno_ess %>% 
  group_by(gene_name) %>% 
  filter(length(gene_name)==2) %>% 
  pull(gene_name) %>% 
  unique()

length(anno_ess_2UTR_list)

# Create new data frame by filtering for the genes in "anno_ess_2UTR_list" and estimate isoform expression levels from total gene expression (after considering ipa) and wui. The results were summarized on the level of the perturbation cluster (means).

nontarg_ess_iso <- df_WUI_ess_clus %>% 
  filter(target_gene=="non-targeting", 
         gene_name %in% anno_ess_2UTR_list) %>% 
  mutate(pct_ipa = replace_na(pct_ipa, 0)) %>%
  group_by(gene_name, gene_id) %>%
  summarise(mean_wui = mean(wui, na.rm=TRUE), 
            mean_tpm = mean(tpm, na.rm=TRUE), 
            mean_ipa = mean(pct_ipa, na.rm=TRUE),
            mean_SU_tpm=mean((1-pct_ipa)*tpm*(1-wui), na.rm=TRUE),
            mean_LU_tpm=mean((1-pct_ipa)*tpm*wui, na.rm=TRUE))


df_WUI_ess_clus_iso <- df_WUI_ess_clus %>% 
  filter(gene_name %in% anno_ess_2UTR_list) %>% 
  mutate(pct_ipa = replace_na(pct_ipa, 0)) %>%
  group_by(gene_name, gene_id, cluster_id_target) %>%
  summarise(wui = mean(wui, na.rm=TRUE), 
            tpm = mean(tpm,na.rm=TRUE), 
            ipa = mean(pct_ipa,na.rm=TRUE),
            SU_tpm=mean((1-pct_ipa)*tpm*(1-wui), na.rm=TRUE),
            LU_tpm=mean((1-pct_ipa)*tpm*wui, na.rm=TRUE))


df_WUI_ess_clus_iso1 <- left_join(df_WUI_ess_clus_iso, nontarg_ess_iso, 
                                  by=c("gene_id", "gene_name")) %>%
  mutate(fc_tpm = log2(tpm/mean_tpm),
         fc_SU_tpm = log2(SU_tpm/mean_SU_tpm),
         fc_LU_tpm = log2(LU_tpm/mean_LU_tpm),
         dSU_tpm = SU_tpm- mean_SU_tpm,
         dLU_tpm = LU_tpm- mean_LU_tpm)

### now summarize to the cluster level

pert_clus_iso <- df_WUI_ess_clus_iso1 %>% group_by(gene_name, gene_id, cluster_id_target) %>% 
  summarize(fc_tpm = mean(fc_tpm, na.rm=T),
            fc_SU_tpm = mean(fc_SU_tpm, na.rm=T),
            fc_LU_tpm =  mean(fc_LU_tpm, na.rm=T),
            dSU_tpm = mean(dSU_tpm, na.rm=T),
            dLU_tpm =  mean(dLU_tpm, na.rm=T)
            ) %>%
  ungroup()

### 1. combine with table of genes that contains only significant APA results
### 2. define different Regulation categories
### 3. summary plot 

### Note: We need to add an extra row for cluster 16 for the lengthening set in order to have two equally-sized data frames.

pert_clus_iso_Len <- pert_clus_iso %>% 
  inner_join(df_WUI_ess_clus_sum2, by=c("gene_name", "gene_id", "cluster_id_target")) %>% filter(!is.na(cluster_id_target)) %>% 
  filter(DUL=="Lengthening") %>% 
  group_by(cluster_id_target) %>%
  summarize("Lengthening w/ compensation"= sum(dLU_tpm/dSU_tpm <= -0.5 & dLU_tpm/dSU_tpm >= -2),
            "Lengthening w/o compensation"= sum(dLU_tpm/dSU_tpm > -0.5 | dLU_tpm/dSU_tpm <  -2),
            "Lengthening w/o compensation SU down only"= sum(dLU_tpm/dSU_tpm > -0.5),
            "Lengthening w/o compensation LU up only"= sum(dLU_tpm/dSU_tpm < -2)) %>%
  add_row(cluster_id_target="16", 
          "Lengthening w/ compensation" =0, 
          "Lengthening w/o compensation" =0,
          "Lengthening w/o compensation SU down only"=0, 
          "Lengthening w/o compensation LU up only"=0)

pert_clus_iso_Sho <-  pert_clus_iso %>% inner_join(df_WUI_ess_clus_sum2, by=c("gene_name", "gene_id", "cluster_id_target")) %>% filter(!is.na(cluster_id_target)) %>% 
  filter(DUL=="Shortening") %>% 
  group_by(cluster_id_target) %>%
  summarize("Shortening w/ compensation"= sum(dSU_tpm/dLU_tpm <= -0.5 & dSU_tpm/dLU_tpm >= -2),
            "Shortening w/o compensation"= sum(dSU_tpm/dLU_tpm > -0.5 | dSU_tpm/dLU_tpm < -2),
            "Shortening w/o compensation SU up only"= sum(dSU_tpm/dLU_tpm > -0.5),
            "Shortening w/o compensation LU down only"= sum(dSU_tpm/dLU_tpm < -2))

pert_clus_iso1 <- left_join(pert_clus_iso_Len, pert_clus_iso_Sho, by ="cluster_id_target") %>%
  mutate("Compensation" = `Lengthening w/ compensation`+`Shortening w/ compensation`,
         "No compensation"= `Lengthening w/o compensation`+ `Shortening w/o compensation`,
         "Compensation_ratio" =  `Compensation`/(`Compensation`+`No compensation`)) %>%
  left_join(names, by="cluster_id_target") %>%
  mutate("SU regulation"= `Lengthening w/o compensation SU down only` + `Shortening w/o compensation SU up only`,
         "LU regulation" = `Lengthening w/o compensation LU up only` + `Shortening w/o compensation LU down only`,
         "Balanced regulation" = `Lengthening w/ compensation`+ `Shortening w/ compensation`)

pert_clus_iso2 <- pert_clus_iso1 %>%
  pivot_longer(
    cols = c("SU regulation",
             "LU regulation",
             "Balanced regulation"),
    names_to = "Regulation", 
    values_to = "Genes") %>%
  mutate(Regulation = factor(Regulation, levels= c("LU regulation","Balanced regulation","SU regulation")))

compensation_plot <- pert_clus_iso2 %>%  
  ggplot(aes(y= reorder(clus_name, -order_heatm_new), 
             x=Genes , fill= Regulation, label=abs(Genes))) +
  geom_col(pos="fill", col="black") +
  theme_bw() +
  ylab("") +
  scale_fill_manual(values = c("#1c75bc","#bcbec0","#609e9f")) +
  labs(fill="Gene regulation") +
  scale_x_continuous(labels = scales::percent, expand=c(0,0))

compensation_plot

ggsave("Figure_outputs/Fig5d.pdf", compensation_plot , width= 12, height=8)

```


Finally, we provide an overview figure showing the gene features that have statistically the strongest association with isoform expression regulation across all clusters. The input data for this figure can be found in Supplementary Data 9.

```{r}

features <- read_xlsx("table_outputs/Suppl_Data9.xlsx", sheet="Sheet1")

feature_analysis <- features %>%
    mutate(Result = case_when(FDR>0.05 ~ "Unaffected",
            FDR<0.05 & sign(estimate)== 1 ~ "Promotes Lengthening",
            FDR<0.05 & sign(estimate)== -1 ~ "Promotes Shortening"),
  Result = factor(Result, levels=c("Promotes Lengthening", "Unaffected", "Promotes Shortening"))) %>% 
  dplyr::count(feature, Result, .drop=FALSE) %>% 
  group_by(feature) %>%
  mutate(order= n[Result == "Promotes Shortening"]*(-1) + n[Result =="Promotes Lengthening"])

overview_features <- feature_analysis %>%
  dplyr::filter(Result %in% c("Promotes Lengthening", "Promotes Shortening")) %>%
  group_by(feature) %>%
  summarize(total = sum(n)) %>%
  ggplot(aes(x= reorder(feature, -total), y=total))+
  scale_y_continuous("Number of perturbation clusters", 
                     limits=c(0,18), 
                     breaks=c(0,3,6,9,12,15,18),
                     expand=c(0,0))+
  geom_col(fill="#bdbdda", col="black")+
  scale_x_discrete(guide = guide_axis(angle = 90)) +
  theme_bw()

overview_features

ggsave("Figure_outputs/Fig5e.pdf", overview_features)

```


---

# Runtime Details
## Session Info
<details>
```{r sesh_info, echo=FALSE}
sessionInfo()
```
</details>

## Conda Environment
<details>
```{bash comment="", echo=FALSE}
if ! command -v conda &> /dev/null
then
  echo "Conda not detected."
elif [ -z "${CONDA_PREFIX}" ]
then
  echo "No active Conda environment."
else
  echo "## Conda Environment YAML"
  conda env export
fi
```
</details>


