---
title: "Fansler et al., Nature Communications, Supplementary Data 6"
author: "Sibylle Mitschka"
output: 
  html_document: 
    code_folding: show
    df_print: paged
---

Loading of R packages (not all may be required for this code section). 


```{r message=FALSE, warning=FALSE}

library(tidyverse)
library(broom)
library(Biostrings)
library(ggpubr)
library(rtracklayer)
library(org.Hs.eg.db)
library(RColorBrewer)
library(biomaRt)
library(GenomicFeatures)
library(BSgenome)
library(GeneOverlap)
library(pheatmap)
library(biomaRt)
library(plyranges)
library(ggrepel)
library(writexl)
library(gridExtra)
library(readxl)

```

## Introduction

This part of the code generates some of the figures and data related to the analsyis of the Perturb-Seq data set presented in figures 4 and 5 and Supplemental figures 4 and 5.

The data for this analysis have been generated by MMF through isoform quantification with the scUTRquant tool based using the "human UTRome". The primary data are from the essential gene CRISPRi screen performed in human K562 cells (Replogle et al Cell 2022). In total, this screen comprised 2.134 targeting perturbations and 97 non-targeting control sgRNA samples. A clustering of z-scaled dWUI values using a subset of response genes previously generated a total of 18 perturbation clusters comprising 836 genes.

In this analysis, we generated statistics describing the effect of these perturbation clusters with respect to their averaged global effects on APA and IPA (Fig S4), the number of regulated genes for APA, their effect direction (lengtheneing vs shortening) and effect strength using the dWUI metric (Fig 5a). 
We performed pairwise comparisons of the sets of regulated genes in order to explore cooperative or antagonistic regulation of multi-UTR genes by different clusters (Fig 5b and 5c). 
Furthermore, we determined whether APA isoform changes are balanced (compensated), or not (Fig 5d). 
Finally, using the cluster-averaged effects on multi-UTR genes, we performed correlation analyses against gene features derived from data generated and outside published data sources and annotations (Fig 5e and 5f).

```{r message=FALSE, warning=FALSE}

### this data file comprises the gene expression counts and WUI and pct_ipa values for all sgRNA and gene pair conditions

df_WUIess <-readRDS("input_data/df_wui_kd6_essential_ui10.Rds")

### This file provides the cluster identity for a subset of the perturbations (836 genes) which will be used to summarize the data on the cluster level, the clsuter_ids provided in this file do not match the final cluster numbers presented in the paper

target_clus <- read.csv("input_data/df_target_nnclusters_kd6_essential_ui10.csv") %>% mutate(cluster_id_target= as.factor(cluster_id))

### This file provides the names for the clusters and the conversion of the primary cluster id to the final cluster id used in the paper.

names <- read.csv("input_data/names_perturbation_cluster.csv") %>% mutate(cluster_id_target= as.factor(cluster_id_target))

### This file provides additional data on the subset of mRNA 3'ends that have been detected in this K562 data set. They will be used to derive further features such as mRNA lengths from the data.

anno_ess <- read.csv("input_data/df_utrs_kd6_essential_ui10.csv")

```


First, we calculate the baseline mean for all genes' WUI, IPA (named pct_ipa) and expression values from the non-targeting control sgRNA conditions in the data set (n=97). 
Mean_wui and mean_ipa were replaced with NA for genes with a mean_tpm <5 to reduce noise in the data. Similar to 3'UTR APA, a gene's IPA was only considered when it ranged between 10 and 90%.
Next, these values were used to calculate the relative deviations in these metrics in the perturbed conditions, i.e. dwui, fc_tpm and dipa. 


```{r message=FALSE, warning=FALSE, results='hide'}

nontarg_ess <- df_WUIess %>% 
  filter(target_gene=="non-targeting") %>% 
  group_by(gene_name, gene_id) %>%
  summarize(mean_tpm = mean(tpm, na.rm=TRUE),
            sd_tpm = sd(tpm, na.rm=TRUE),
            mean_wui = mean(wui, na.rm=TRUE),
            sd_wui= sd(wui, na.rm=TRUE), 
            mean_ipa = mean(pct_ipa, na.rm=TRUE),
            sd_ipa= sd(pct_ipa, na.rm=TRUE)) %>%
  mutate(mean_wui = if_else(mean_tpm>= 5, mean_wui, NA),
         mean_ipa = if_else(mean_tpm>= 5 & mean_ipa>=0.1 & mean_ipa<= 0.9 , mean_ipa, NA),
         sd_wui= if_else(mean_tpm>= 5,sd_wui, NA),
         sd_ipa= if_else(mean_tpm>= 5 & mean_ipa>=0.1 & mean_ipa<= 0.9 ,sd_ipa, NA))

            
df_WUIess1 <- left_join(df_WUIess, 
                        nontarg_ess, by=c("gene_id", "gene_name"), 
                        multiple="all") %>%
              mutate(dwui = wui-mean_wui, 
                     dipa = pct_ipa-mean_ipa,
                     fc_tpm = log2(tpm/mean_tpm),
                     z_dwui= dwui/sd_wui,
                    z_dipa= dipa/sd_ipa,
                    z_tpm= (tpm-mean_tpm)/sd_tpm)

```

Here we are providing a summary of global average deviation of WUI and IPA for each perturbation, called a_dwui and a_dipa, respectively. The output of this summary was provided as a supplementary table.

```{r}

a_dwui <- df_WUIess1 %>% 
  group_by(target_gene, target_gene_id, sgID_AB, n_cells) %>% 
  summarize(a_dwui = mean(dwui, na.rm=TRUE)) %>% ungroup()

a_dipa <- df_WUIess1 %>% 
  group_by(target_gene, target_gene_id, sgID_AB) %>%   
  summarize(a_dipa = mean(dipa, na.rm=TRUE)) %>% ungroup()

a_dwui_dipa <- left_join(a_dwui, a_dipa, by= c("target_gene", "target_gene_id", "sgID_AB"))

sum_clus <- left_join(a_dwui_dipa, target_clus, by= c("target_gene", "sgID_AB", "target_gene_id")) %>%
  mutate(cluster_id_target = as.factor(cluster_id)) %>% 
  dplyr:: select(-cluster_id) %>%
  left_join(names, by="cluster_id_target") %>% 
  mutate(cluster_id_target= ifelse(target_gene=="non-targeting", "control", cluster_id_target)) %>%
  select(target_gene, target_gene_id, sgID_AB, n_cells, order_heatm_new, clus_name, a_dwui, a_dipa)

sum_clus1 <- sum_clus %>% 
  dplyr::rename(Gene_symbol =	target_gene,
                Ensembl_ID  = target_gene_id,
                Cluster_id  = order_heatm_new,
                Cluster_name = clus_name)

write_xlsx(sum_clus1, "table_outputs/Supplementary_Data_6.xlsx")

```


---

# Runtime Details
## Session Info
<details>
```{r sesh_info, echo=FALSE}
sessionInfo()
```
</details>

## Conda Environment
<details>
```{bash comment="", echo=FALSE}
if ! command -v conda &> /dev/null
then
  echo "Conda not detected."
elif [ -z "${CONDA_PREFIX}" ]
then
  echo "No active Conda environment."
else
  echo "## Conda Environment YAML"
  conda env export
fi
```
</details>

